File: vps-telegram-bot/vpsupload_bot/video_bot_mod/media_utils.py
"""媒体文件、路径、安全与格式化工具。"""

import asyncio
import logging
import math
import os
import re
from datetime import datetime
from typing import Dict, List, Optional, Tuple

from .config import BASE_DIR, FFPROBE_TIMEOUT_SECONDS, MERGE_MIN_DURATION_RATIO, MERGE_MIN_SIZE_RATIO

logger = logging.getLogger(__name__)

def assert_path_inside_base(path: str) -> str:
    """确保所有文件操作都限制在 BASE_DIR 内，避免状态异常导致路径越界。"""
    base = os.path.realpath(BASE_DIR)
    target = os.path.realpath(path)
    if target == base or target.startswith(base + os.sep):
        return target
    raise ValueError(f"非法路径，已超出 BASE_DIR: {target}")

def safe_join(parent: str, child: str) -> str:
    return assert_path_inside_base(os.path.join(parent, child))

def remove_if_exists(path: str) -> None:
    try:
        if path and os.path.exists(path):
            os.remove(path)
    except OSError as exc:
        logger.warning("清理文件失败: %s, %s", path, exc)

def unique_path(path: str) -> str:
    """如果目标文件已存在，自动生成 xxx_1.ext、xxx_2.ext，避免 ffmpeg 覆盖旧文件。"""
    path = os.path.abspath(path)
    if not os.path.exists(path):
        return path

    root, ext = os.path.splitext(path)
    index = 1
    while True:
        candidate = f"{root}_{index}{ext}"
        if not os.path.exists(candidate):
            return candidate
        index += 1

def get_formatted_file_size(filepath: str) -> str:
    """智能获取文件大小：小于 1GB 显示 MB，否则显示 GB。"""
    try:
        size_bytes = os.path.getsize(filepath)
        size_mb = size_bytes / (1024 ** 2)
        if size_mb >= 1024:
            size_gb = size_bytes / (1024 ** 3)
            return f"{size_gb:.2f}GB"
        return f"{size_mb:.2f}MB"
    except OSError:
        return "0.00MB"

def format_duration(seconds: float) -> str:
    if seconds <= 0:
        return "未知"
    seconds_int = int(seconds)
    h = seconds_int // 3600
    m = (seconds_int % 3600) // 60
    s = seconds_int % 60
    return f"{h:02d}:{m:02d}:{s:02d}"

async def get_video_duration(filepath: str) -> float:
    """Use ffprobe to read media duration. Return 0.0 and log details on failure."""
    cmd = [
        "ffprobe",
        "-v", "error",
        "-show_entries", "format=duration",
        "-of", "default=noprint_wrappers=1:nokey=1",
        filepath,
    ]
    process = None
    try:
        process = await asyncio.create_subprocess_exec(
            *cmd,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await asyncio.wait_for(
            process.communicate(),
            timeout=FFPROBE_TIMEOUT_SECONDS,
        )
    except asyncio.TimeoutError:
        if process and process.returncode is None:
            process.kill()
            await process.communicate()
        logger.warning("ffprobe 超时: file=%s timeout=%ss", filepath, FFPROBE_TIMEOUT_SECONDS)
        return 0.0
    except FileNotFoundError:
        logger.exception("ffprobe 未安装或不可执行，无法分析文件时长: %s", filepath)
        return 0.0
    except Exception as exc:
        logger.exception("ffprobe 执行异常: file=%s error=%s", filepath, exc)
        return 0.0

    stderr_text = stderr.decode("utf-8", errors="ignore").strip()
    stdout_text = stdout.decode("utf-8", errors="ignore").strip()

    if process.returncode != 0:
        logger.warning(
            "ffprobe 分析失败: file=%s returncode=%s stderr=%s",
            filepath,
            process.returncode,
            stderr_text[:500],
        )
        return 0.0

    try:
        return float(stdout_text)
    except ValueError:
        logger.warning(
            "ffprobe 输出无法解析为时长: file=%s stdout=%r stderr=%s",
            filepath,
            stdout_text,
            stderr_text[:500],
        )
        return 0.0

def build_progress_bar(percent: float, length: int = 20) -> str:
    percent = max(0.0, min(100.0, percent))
    filled = int(math.floor((percent / 100.0) * length))
    bar = "█" * filled + "░" * (length - filled)
    return f"[{bar}] {percent:5.1f}%"

def smart_rename(first_file_path: str) -> str:
    base_name = os.path.splitext(os.path.basename(first_file_path))[0]
    ext = os.path.splitext(first_file_path)[1]

    # 增强正则：匹配日期(YYYY-MM-DD) 以及其后的时间(HH-MM-SS)，保留完整的录播时间标签
    date_match = re.search(r"\d{4}[-_.]\d{2}[-_.]\d{2}(?:[ _-]\d{2}[-_.:]\d{2}[-_.:]\d{2})?", base_name)
    date_str = date_match.group(0) if date_match else f"Merged_{datetime.now().strftime('%Y%m%d_%H%M%S')}"

    # 清理掉开头的日期时间括号 [2026-06-07 20-00-00]，保留后面的标题
    title_part = re.sub(r"^\[\d[^\]]*\]\s*", "", base_name)

    if title_part == base_name:
        title_part = base_name.replace(date_str, "")
        title_part = re.sub(r"^[-_.\s]+", "", title_part)

    if title_part:
        output_name = f"{date_str}_{title_part}{ext}"
    else:
        output_name = f"{date_str}_merged{ext}"

    return output_name.replace("__", "_")

def write_concat_list(list_file_path: str, file_paths: List[str]) -> None:
    """写入 ffmpeg concat list，并转义单引号和反斜杠。"""
    with open(list_file_path, "w", encoding="utf-8") as list_file:
        for file_path in file_paths:
            escaped = file_path.replace("\\", "\\\\").replace("'", "\\'")
            list_file.write(f"file '{escaped}'\n")

async def validate_merged_file(
    output_path: str,
    returncode: Optional[int],
    input_total_duration: float,
    input_total_size: int,
) -> Tuple[bool, Dict[str, float]]:
    """合并成功判断：ffmpeg 退出码 + 输出存在 + 时长达标 + 大小辅助校验。"""
    output_size = os.path.getsize(output_path) if os.path.exists(output_path) else 0
    output_duration = await get_video_duration(output_path) if output_size > 0 else 0.0

    duration_ok = True
    if input_total_duration > 0:
        duration_ok = output_duration >= input_total_duration * MERGE_MIN_DURATION_RATIO

    size_ok = True
    if input_total_size > 0:
        size_ok = output_size >= input_total_size * MERGE_MIN_SIZE_RATIO

    is_success = bool(returncode == 0 and output_size > 0 and duration_ok and size_ok)
    details = {
        "input_duration": input_total_duration,
        "output_duration": output_duration,
        "input_size": float(input_total_size),
        "output_size": float(output_size),
        "duration_ratio": (output_duration / input_total_duration) if input_total_duration > 0 else 0.0,
        "size_ratio": (output_size / input_total_size) if input_total_size > 0 else 0.0,
        "duration_ok": 1.0 if duration_ok else 0.0,
        "size_ok": 1.0 if size_ok else 0.0,
    }
    return is_success, details

def format_merge_check(details: Dict[str, float]) -> str:
    input_duration = details.get("input_duration", 0.0)
    output_duration = details.get("output_duration", 0.0)
    duration_ratio = details.get("duration_ratio", 0.0) * 100
    size_ratio = details.get("size_ratio", 0.0) * 100
    output_size = int(details.get("output_size", 0.0))

    return (
        f"⏱️ 输入总时长: `{format_duration(input_duration)}`\n"
        f"⏱️ 输出时长: `{format_duration(output_duration)}` ({duration_ratio:.1f}%)\n"
        f"📦 输出大小: `{output_size / (1024 ** 2):.2f}MB` ({size_ratio:.1f}%)"
    )

def format_elapsed(seconds: float) -> str:
    seconds_int = max(0, int(seconds))
    h = seconds_int // 3600
    m = (seconds_int % 3600) // 60
    s = seconds_int % 60
    if h:
        return f"{h}h{m:02d}m{s:02d}s"
    if m:
        return f"{m}m{s:02d}s"
    return f"{s}s"

File: vps-telegram-bot/vpsupload_bot/RUST/src/media_utils.rs
use anyhow::{anyhow, Result};
use chrono::Local;
use regex::Regex;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Stdio;
use tokio::io::{AsyncBufReadExt, BufReader};
use tokio::process::Command;
use tokio::time::{timeout, Duration};

pub fn assert_path_inside_base(base: &Path, target: &Path) -> Result<PathBuf> {
    let base_canon = base.canonicalize().unwrap_or_else(|_| base.to_path_buf());
    let target_canon = if target.exists() {
        target.canonicalize()?
    } else {
        target.to_path_buf()
    };

    if target_canon == base_canon || target_canon.starts_with(&base_canon) {
        Ok(target_canon)
    } else {
        Err(anyhow!("非法路径，已超出 BASE_DIR: {:?}", target))
    }
}

pub fn safe_join(base: &Path, parent: &Path, child: &str) -> Result<PathBuf> {
    let joined = parent.join(child);
    assert_path_inside_base(base, &joined)
}

pub fn remove_if_exists(path: &Path) {
    if path.exists() {
        let _ = fs::remove_file(path);
    }
}

pub fn unique_path(path: &Path) -> PathBuf {
    if !path.exists() {
        return path.to_path_buf();
    }
    let stem = path.file_stem().and_then(|s| s.to_str()).unwrap_or("file");
    let ext = path.extension().and_then(|s| s.to_str()).unwrap_or("");
    let parent = path.parent().unwrap_or_else(|| Path::new("."));
    let mut index = 1;
    loop {
        let candidate_name = if ext.is_empty() {
            format!("{}_{}", stem, index)
        } else {
            format!("{}_{}.{}", stem, index, ext)
        };
        let candidate = parent.join(candidate_name);
        if !candidate.exists() {
            return candidate;
        }
        index += 1;
    }
}

pub fn get_formatted_file_size(path: &Path) -> String {
    match fs::metadata(path) {
        Ok(meta) => {
            let size_bytes = meta.len() as f64;
            let size_mb = size_bytes / (1024.0 * 1024.0);
            if size_mb >= 1024.0 {
                format!("{:.2}GB", size_bytes / (1024.0 * 1024.0 * 1024.0))
            } else {
                format!("{:.2}MB", size_mb)
            }
        }
        Err(_) => "0.00MB".to_string(),
    }
}

pub fn format_duration(seconds: f64) -> String {
    if seconds <= 0.0 {
        return "未知".to_string();
    }
    let s_int = seconds as u64;
    let h = s_int / 3600;
    let m = (s_int % 3600) / 60;
    let s = s_int % 60;
    format!("{:02}:{:02}:{:02}", h, m, s)
}

pub async fn get_video_duration(path: &Path, timeout_sec: u64) -> f64 {
    let mut cmd = Command::new("ffprobe");
    cmd.args(&[
        "-v", "error",
        "-show_entries", "format=duration",
        "-of", "default=noprint_wrappers=1:nokey=1",
    ])
    .arg(path)
    .stdout(Stdio::piped())
    .stderr(Stdio::piped());

    let child = match cmd.spawn() {
        Ok(c) => c,
        Err(_) => return 0.0,
    };

    match timeout(Duration::from_secs(timeout_sec), child.wait_with_output()).await {
        Ok(Ok(output)) if output.status.success() => {
            let s = String::from_utf8_lossy(&output.stdout).trim().to_string();
            s.parse::<f64>().unwrap_or(0.0)
        }
        _ => 0.0,
    }
}

pub fn build_progress_bar(percent: f64, length: usize) -> String {
    let p = percent.clamp(0.0, 100.0);
    let filled = ((p / 100.0) * (length as f64)).floor() as usize;
    let empty = length.saturating_sub(filled);
    format!("[{}{}] {:5.1}%", "█".repeat(filled), "░".repeat(empty), p)
}

pub fn smart_rename(first_file: &Path) -> String {
    let base_name = first_file.file_stem().and_then(|s| s.to_str()).unwrap_or("Merged");
    let ext = first_file.extension().and_then(|s| s.to_str()).unwrap_or("mp4");

    let date_re = Regex::new(r"\d{4}[-_.]\d{2}[-_.]\d{2}(?:[ _-]\d{2}[-_.:]\d{2}[-_.:]\d{2})?").unwrap();
    let date_str = match date_re.find(base_name) {
        Some(m) => m.as_str().to_string(),
        None => format!("Merged_{}", Local::now().format("%Y%m%d_%H%M%S")),
    };

    let title_clean_re = Regex::new(r"^\[\d[^\]]*\]\s*").unwrap();
    let mut title_part = title_clean_re.replace(base_name, "").to_string();

    if title_part == base_name {
        title_part = title_part.replace(&date_str, "");
        let trim_re = Regex::new(r"^[-_.\s]+").unwrap();
        title_part = trim_re.replace(&title_part, "").to_string();
    }

    let output_name = if !title_part.is_empty() {
        format!("{}_{}.{}", date_str, title_part, ext)
    } else {
        format!("{}_merged.{}", date_str, ext)
    };

    output_name.replace("__", "_")
}

pub fn write_concat_list(list_path: &Path, files: &[PathBuf]) -> Result<()> {
    let mut content = String::new();
    for f in files {
        let s = f.to_str().unwrap_or("").replace('\\', "\\\\").replace('\'', "\\'");
        content.push_str(&format!("file '{}'\n", s));
    }
    fs::write(list_path, content)?;
    Ok(())
}

pub struct MergeCheckResult {
    pub is_success: bool,
    pub input_duration: f64,
    pub output_duration: f64,
    pub duration_ratio: f64,
    pub size_ratio: f64,
    pub output_size: u64,
}

pub async fn validate_merged_file(
    output_path: &Path,
    success: bool,
    input_total_duration: f64,
    input_total_size: u64,
    min_dur_ratio: f64,
    min_size_ratio: f64,
    timeout_sec: u64,
) -> MergeCheckResult {
    let output_size = fs::metadata(output_path).map(|m| m.len()).unwrap_or(0);
    let output_duration = if output_size > 0 {
        get_video_duration(output_path, timeout_sec).await
    } else {
        0.0
    };

    let duration_ok = if input_total_duration > 0.0 {
        output_duration >= input_total_duration * min_dur_ratio
    } else {
        true
    };

    let size_ok = if input_total_size > 0 {
        (output_size as f64) >= (input_total_size as f64) * min_size_ratio
    } else {
        true
    };

    let is_success = success && output_size > 0 && duration_ok && size_ok;
    let duration_ratio = if input_total_duration > 0.0 { output_duration / input_total_duration } else { 0.0 };
    let size_ratio = if input_total_size > 0 { (output_size as f64) / (input_total_size as f64) } else { 0.0 };

    MergeCheckResult {
        is_success,
        input_duration: input_total_duration,
        output_duration,
        duration_ratio,
        size_ratio,
        output_size,
    }
}

pub fn format_merge_check(res: &MergeCheckResult) -> String {
    format!(
        "⏱️ 输入总时长: `{}`\n⏱️ 输出时长: `{}` ({:.1}%)\n📦 输出大小: `{:.2}MB` ({:.1}%)",
        format_duration(res.input_duration),
        format_duration(res.output_duration),
        res.duration_ratio * 100.0,
        (res.output_size as f64) / (1024.0 * 1024.0),
        res.size_ratio * 100.0
    )
}

pub fn format_elapsed(seconds: f64) -> String {
    let s_int = seconds.max(0.0) as u64;
    let h = s_int / 3600;
    let m = (s_int % 3600) / 60;
    let s = s_int % 60;
    if h > 0 { format!("{}h{:02}m{:02}s", h, m, s) }
    else if m > 0 { format!("{}m{:02}s", m, s) }
    else { format!("{}s", s) }
}